File: Python/0988.py
from typing import Optional

#Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def dfs(self, node):
        if not node:
            return

        self.path.append(chr(node.val + ord("a")))

        if not node.left and not node.right:
            cur = "".join(self.path[::-1])
            self.ans = min(cur, self.ans)

        self.dfs(node.left)
        self.dfs(node.right)

        self.path.pop()

    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        self.ans = "{"
        self.path = []
        self.dfs(root)

        return self.ans

# Topic: Tree, DFS
# Time complexity: O(n)
# Space complexity: O(h) where h is the height of the tree

File: C++/0988.cpp
#include <vector>
#include <string>
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::string smallestFromLeaf(TreeNode* root) {
        std::string ans = "{";
        std::vector<char> path;
        dfs(root, path, ans);
        return ans;
    }

private:
    void dfs(TreeNode* node, std::vector<char>& path, std::string& ans) {
        if (!node) {
            return;
        }

        path.push_back(static_cast<char>(node->val + 'a'));

        if (!node->left && !node->right) {
            std::string cur(path.rbegin(), path.rend());
            ans = std::min(ans, cur);
        }

        dfs(node->left, path, ans);
        dfs(node->right, path, ans);

        path.pop_back();
    }
};

// Topic: Tree, Dfs
// Time Complexity: O(n)
// Space Complexity: O(h) height of the tree